source/common/stats/allocator_impl.cc

```diff
@@ -316,5 +336,77 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
   return new CounterImpl(name, *this, tag_extracted_name, stat_name_tags);
 }
 
+void AllocatorImpl::forEachCounter(std::function<void(std::size_t)> f_size,
+                                   std::function<void(Stats::Counter&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  if (f_size != nullptr) {
+    f_size(counters_.size());
+  }
+  for (auto& counter : counters_) {
+    f_stat(*counter);
+  }
+}
+
+void AllocatorImpl::forEachGauge(std::function<void(std::size_t)> f_size,
+                                 std::function<void(Stats::Gauge&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  if (f_size != nullptr) {
+    f_size(gauges_.size());
+  }
+  for (auto& gauge : gauges_) {
+    f_stat(*gauge);
+  }
+}
+
+void AllocatorImpl::forEachTextReadout(std::function<void(std::size_t)> f_size,
+                                       std::function<void(Stats::TextReadout&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  if (f_size != nullptr) {
+    f_size(text_readouts_.size());
+  }
+  for (auto& text_readout : text_readouts_) {
+    f_stat(*text_readout);
+  }
+}
+
+void AllocatorImpl::markCounterForDeletion(const CounterSharedPtr& counter) {
+  Thread::LockGuard lock(mutex_);
+  auto iter = counters_.find(counter->statName());
+  if (iter == counters_.end()) {
+    // This has already been marked for deletion.
+    return;
+  }
+  ASSERT(counter.get() == *iter);
+  // Duplicates are ASSERTed in ~AllocatorImpl.
+  deleted_counters_.emplace_back(*iter);
+  counters_.erase(iter);
```

Comment:
I thought of a concern with this tactic for this class: if someone allocates a counter with the same name as one that was marked for deletion, we'll wind with two different counter objects with the same name. That might be OK but we might want to add a test.

It won't actually happen with ThreadLocalStore because it only calls markCounterForDeletion on stats that are rejected by the matcher, and the matcher can't be changed once it's added. The reason that this is needed is that the matcher can be added after some stats are created, and we need to effectively get rid of the now-rejected stats without causing references to freed memory.

Possible remedies I thought of include:
 * add a test and accept that as a weird use-case for this class, that won't be hit by ThreadLocalStore.
 * prevent this by checking against deleted_counters_ (which could be a set) during allocation
 * declare this invalid by asserting against deleted_counters_ for debug builds
 * use a bit in `flags_` to denote that a stat needs be considered as marked for deletion, and skip that during forEach. You could report the correct size by keeping track of the number of deleted counters in the class. I think this is the biggest change from what you have, but would mean that if you try to allocate a stat with the same name as one that was deleted, you'll get back the previously deleted stat.

Reply:
> use a bit in flags_ to denote that a stat needs be considered as marked for deletion, and skip that during forEach. You could report the correct size by keeping track of the number of deleted counters in the class. I think this is the biggest change from what you have, but would mean that if you try to allocate a stat with the same name as one that was deleted, you'll get back the previously deleted stat.

This approach sounds good to me, fwiw.

Stepping back though, I'm trying to understand why this logic was moved into the allocator versus where it used to be in thread local store? Can you help me understand that?